### ISSUE TITLE
Why are some strings quoted?

### ISSUE THREAD

**Issue by edent** [citation-file-format/citation-file-format#322]
In the examples, some strings are quoted, and some aren't:

```
authors:
  - family-names: Druskat
    given-names: Stephan
    name-particle: von
    name-suffix: III
    address: "742 Evergreen Terrace"
    affiliation: "German Aerospace Center (DLR)"
    alias: "sdruskat"
    city: "Berlin"
    country: "DE"
    email: "sdruskat@research-project.org"
    fax: +12-3456-7890
    orcid: "https://orcid.org/0000-0003-4925-7248"
    post-code: 90210
    region: Renfrewshire
    tel: +12-345-6789098
    website: "https://research-project.org"
  ```
and

```
abstract: This is my awesome research software. It does many things.
authors:
  - family-names: Druskat
    given-names: Stephan
    orcid: "https://orcid.org/0000-0003-4925-7248"
cff-version: 1.2.0
date-released: "2021-07-18"
identifiers:
  - description: This is the collection of archived snapshots of all versions of My Research Software
    type: doi
    value: 10.5281/zenodo.123456
  - description: This is the archived snapshot of version 0.11.2 of My Research Software
    type: doi
    value: 10.5281/zenodo.123457
keywords:
  - "awesome software"
  - research
license: Apache-2.0
message: If you use this software, please cite it using these metadata.
repository-code: "https://github.com/citation-file-format/my-research-software"
title: My Research Software
version: 0.11.2
```
What's the logic behind which strings are quoted and which aren't? Some strings with spaces are quoted, some aren't. `date-released:` sometimes has them, sometimes not.

Is there a simple guide to which ones need to be quoted? Would it make sense to wrap *all* values in quotes in the examples?

**Comment by sdruskat** [comment]
Hi, @edent, and thanks for opening this issue!

> TL;DR: For most values it doesn't matter whether you quote them or not, as YAML will treat them as strings unless they conform to another data type or contain reserved characters.

You're right, this may be confusing. There are two points here:

1. The format that CFF uses, [YAML](http://yaml.org/), usually treats things as strings, as long as they don't fit another data type (e.g., a number, or a date/timestamp). Quotes force a value to be interpreted as a string. Therefore, while it is arguably *safer* to quote everything that should be a string, even more complex examples, such as `10.5281/zenodo.123456` are treated as a string. This means, you can leave out the quotes to save you some typing.
2. The dates that are used in CFF (always in `YYYY-MM-DD` format, e.g., `2021-08-23`) conform to the YAML `timestamp` data type. However, as the language we provide the [schema](https://github.com/citation-file-format/citation-file-format/blob/main/schema.json) for validation in, [JSON Schema](https://json-schema.org/), doesn't "know" these raw datatypes, you have to convert the dates to a string if you write a validation script for example, so that you can use the JSON Schema to validate this against a specific string pattern (basically, `YYYY-MM-DD`). Hence, it doesn't really matter whether you provide dates in CFF quoted (as string, truly valid CFF) or unquoted (as timestamp, forcibly being made valid by any tooling that uses the schema). This is [documented in a comment in the schema](https://github.com/citation-file-format/citation-file-format/blob/main/schema.json#L282-L290).

In general, we had decided to not include this information in the CFF documentation/schema guide, because we thought it may be too confusing for users not familiar with or interested in things like datatypes, and instead link to the YAML docs where this is documented, if not very obvisouly. But perhaps we should document his after all (point in case, @jspaaks)?
What would have helped you, @edent, to avoid the confusion stemming from the mixed-form examples?

**Comment by edent** [comment]
OK, that makes sense. I think it would be helpful if all the strings were quoted in examples. That way there's no ambiguity.

> The format that CFF uses, YAML, *usually* treats things as strings, 

Sure, it's slightly more typing - but I guess most people don't have an innate understanding of the vagaries of YAML parsing.

Assuming all the strings can be quoted, they should be in the examples. It means people like me don't get stuck wondering where it is appropriate to use them.

Thanks for the swift response - and the brilliant .cff :-)

**Comment by acli** [comment]
I’ve spent a bit of time figuring out the answer to the same question. What I found is that if it can confuse YAML, you need to quote it. This basically means if the value has a colon, you *must* quote it. This is why URL’s are always quoted.

I suspect the same holds if the value starts with a hyphen. I’m not sufficiently familiar with YAML to say if this is indeed the case. I don’t believe a string that starts with a number (e.g., addresses) will confuse YAML.

**Comment by jspaaks** [comment]
> Assuming all the strings can be quoted, they should be in the examples. It means people like me don't get stuck wondering where it is appropriate to use them.

I think I agree with this, people who don't know YAML or don't feel like looking up the specifics every time can just quote everything, knowing that it'll work, while people who are more experienced / more particular can read an example and omit the quotes while understanding the consequences that may have. 

For 1.2.0, we made sure that any key whose value could be interpreted as a number (e.g. `month`, `issue`, `post-code`), is also valid when quoted. See for example:

https://github.com/citation-file-format/citation-file-format/blob/396f738fb025b1d8acdb02a56ffc923f95dc8999/schema.json#L1071-L1082

**Comment by sdruskat** [comment]
Additionally, there should be an example for each of the values with mixed types (any of `number` or `string`) in the definition in the schema guide. In addition to that, I agree, let's quote all values for "pure" string fields in the guide to avoid confusion.

**Comment by acli** [comment]
I respectfully disagree. YAML is supposed to be “human friendly” (quoting from the front page of the YAML website). Quoting everything makes it look like code – not human-friendly.

I spent those 3 hours in that rabbit hole because

1. let’s face it, YAML is not a well-known language (until I encountered CFF I’d never had to deal with it) so I was just extrapolating from the examples
2. there’s no explanation of YAML syntax in our schema guide
3. my first interactions with CFF was through GitHub (this is still the case), GitHub’s error messages are worse than useless
4. I would have tried to validate my CFF, except none of the validators even run on my system and there are no online validators

I had to read the actual schema.json file and went to the YAML website and read the standard to figure out CFF syntax. I strongly believe if the guide even explained CFF syntax (YAML syntax tailored to CFF’s specific circumstances) a little more it would reduce a lot of confusion.

We also need a good online validator, I believe, and by good I mean the error messages should make sense to someone whose first encounter with YAML is CFF.

**Comment by sdruskat** [comment]
> I had to read the actual schema.json file and went to the YAML website and read the standard to figure out CFF syntax. I strongly believe if the guide even explained CFF syntax (YAML syntax tailored to CFF’s specific circumstances) a little more it would reduce a lot of confusion.

I see. In 1.0.3, we had a bit about formatting in the equivalent of the schema guide. It said

> ## Formatting
>
> CFF follows the formatting rules of YAML 1.2, of which one of
> the most important ones is that the colon (`:`) after a key should always be
> followed by a whitespace.
>
> Structure is determined by indentation, i.e., lines
> containing nested elements must be indented by at least one whitespace character,
> although using at least two whitespaces as a standard for indentation preserves
> readability.
>
> Value strings can (and sometimes should) be double-quoted, e.g. `"string"`,
> especially when they contain YAML special characters, or special characters in
> general. These include:
>
> `:  {  }  [  ]  ,  &  *  #  ?  |  -  <  >  =  !  %  @  \`

Would something similar be useful?
This would be an alternative route to quoting everything, and I think I'd actually prefer that.

We had originally removed this from the schema guide as it is documented elsewhere (YAML docs),
and we didn't want to duplicate existing information,
but it seems to be something that would increase user-friendliness of the resources.

> We also need a good online validator, I believe, and by good I mean the error messages should make sense to someone whose first encounter with YAML is CFF.

Yes, we do, and we're trying to get this work funded as well. (Similar to the new version of the initializer website.)
